source/HardwareSerial.cpp

```diff
@@ -2,28 +2,60 @@
 // Licensed under the BSD 2-Clause License.  
 // See License.txt in the project root for license information.
 
+//#include <ppltasks.h>
 #include "arduino.h"
+//#include <concrt.h>   // TODO: Remove.
+
 #ifndef USE_NETWORKSERIAL
 
+#if !WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)   // If building a UWP app:
+#define FORMAT_MESSAGE_ALLOCATE_BUFFER 0x00000100       // From WinBase.h
+
+using namespace Windows::Devices::Enumeration;
+using namespace Windows::Devices::Custom;
+using namespace Windows::Foundation;
+using namespace Windows::Foundation::Collections;
+using namespace Windows::Storage::Streams;
+using namespace Windows::System::Threading;
+using namespace Concurrency;
+
+#define SERIAL_READ_LENGTH 1024
+#define REQUEST_TIMED_OUT -1
+#define NO_PEEK_BYTE REQUEST_TIMED_OUT
+
+#endif // !WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)
+
 void LogLastError()
 {
     CHAR *msg;
     DWORD errCode = GetLastError();
     FormatMessageA(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS, NULL, errCode, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), (LPSTR)&msg, 0, NULL);
     Log(msg);
+
+#if !WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)   // If building a UWP app:
+    HeapFree(GetProcessHeap(), 0, msg);
+#endif // !WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)
+
+#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)   // If building a Win32 app:
+    LocalFree(msg);
+#endif // WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)
 }
 
-HardwareSerial::HardwareSerial(const std::wstring &comPort)
+/// Constructor.
+HardwareSerial::HardwareSerial()
     :
-    _comHandle(INVALID_HANDLE_VALUE),
-    _comPortName(comPort),
-    _storageCount(0),
-    _storageIndex(0),
-    _storageUsed(false),
-    _timeout(1000)
+    m_currentReadNumber(0),
+    m_queuedReadCount(0),
+    m_timeout(1000),
+    m_peekByte(NO_PEEK_BYTE),
+    m_reader(nullptr),
+    m_readThreadCount(0),
+    m_cancellationTokenSource(nullptr)
 {
+    InitializeCriticalSection(&m_readBufferListLock);
```

Comment:
Since this can throw, should it be InitializeCriticalSectionAndSpinCount? Or should we allow the throw